includes/sanitizers/class-amp-core-theme-sanitizer.php

```diff
@@ -1522,4 +1569,446 @@ public function add_twentyfourteen_search() {
+	 */
+	protected function get_element_id( $element, $prefix = 'amp-wp-id' ) {
+		static $index_counter = [];
+
+		if ( $element->hasAttribute( 'id' ) ) {
+			return $element->getAttribute( 'id' );
+		}
+
+		if ( ! array_key_exists( $prefix, $index_counter ) ) {
+			$index_counter[ $prefix ] = 2;
+			$element->setAttribute( 'id', $prefix );
+
+			return $prefix;
+		}
+
+		$id = "{$prefix}-{$index_counter[ $prefix ]}";
+		$index_counter[ $prefix ] ++;
+
+		$element->setAttribute( 'id', $id );
+
+		return $id;
+	}
+
+	/**
+	 * Provides a "best guess" as to what XPath would mirror a given CSS
+	 * selector.
+	 *
+	 * This is a very simplistic conversion and will only work for very basic
+	 * CSS selectors. Therefore, it provides a filter for themes & plugins to
+	 * hook into to provide custom replacements.
+	 *
+	 * @param string $css_selector CSS selector to convert.
+	 * @return string|null XPath that closely mirrors the provided CSS
+	 *                     selector, or null if an error occurred.
+	 */
+	protected function xpath_from_css_selector( $css_selector ) {
+		// Start with basic clean-up.
+		$css_selector = trim( $css_selector );
+		$css_selector = preg_replace( '/\s+/', ' ', $css_selector );
+
+		$xpath             = '';
+		$direct_descendant = false;
+		$token             = strtok( $css_selector, ' ' );
+
+		while ( false !== $token ) {
+			$matches = [];
+
+			// Direct descendant.
+			if ( preg_match( '/^>$/', $token, $matches ) ) {
+				$direct_descendant = true;
+				$token             = strtok( ' ' );
+				continue;
+			}
+
+			// Single class.
+			if ( preg_match( '/^\.(?<class>[a-zA-Z0-9-_]*)$/', $token, $matches ) ) {
+				$descendant        = $direct_descendant ? '/' : '//';
+				$xpath            .= "{$descendant}*[ @class and contains( concat( ' ', normalize-space( @class ), ' ' ), ' {$matches['class']} ' ) ]";
+				$direct_descendant = false;
+				$token             = strtok( ' ' );
+				continue;
+			}
+
+			// Element.
+			if ( preg_match( '/^(?<element>[^.][a-zA-Z0-9-_]*)$/', $token, $matches ) ) {
+				$descendant        = $direct_descendant ? '/' : '//';
+				$xpath            .= "{$descendant}{$matches['element']}";
+				$direct_descendant = false;
+				$token             = strtok( ' ' );
+				continue;
+			}
+
+			$token = strtok( ' ' );
+		}
+
+		return $xpath;
+	}
+
+	/**
+	 * Register an AMP action to an event on a given element.
+	 *
+	 * If the element already contains one or more events or actions, the method
+	 * will assemble them in a smart way.
+	 *
+	 * @param DOMElement $element Element to add an action to.
+	 * @param string     $event   Event to trigger the action on.
+	 * @param string     $action  Action to add.
+	 */
+	protected function add_amp_action( DOMElement $element, $event, $action ) {
+		if ( ! $element->hasAttribute( 'on' ) ) {
+			$element->setAttribute( 'on', "{$event}:{$action}" );
+			return;
+		}
+
+		$events = explode( ';', $element->getAttribute( 'on' ) );
```

Comment:
I think this will be somewhat brittle because it's possible for `;` to appear inside of a method argument, though granted this is unlikely. Instead of exploding perhaps it would be better to inject `$action` into the attribute value right after matching `"$event:"`? Or if it doesn't exist, then just amend `";$event:$action"`? Since this doesn't have a proper parser of the action syntax, I'm wary of giving a false sense of security about using it generally.

In this case perhaps it is better not to include in `AMP_DOM_Utils`.

Reply:
How about making it more robust then? The further we progress on "automating" the transformation into AMP code, the more we'll need to rely on events and actions.

Reply:
OK, please make it more robust then.

Reply:
@westonruter: Can you spot an issue with the following regex: https://regex101.com/r/PvNt9y/1 ?

I think this should be set up so that it only looks after the important bits and completely ignores anything within parentheses.

Reply:
The problematic case you mentioned above is handled and tested.